feat: create kms key as part of cluster bootstrap

aramase opened the pull request:
<!-- Thank you for helping AKS Engine with a pull request!
Use conventional commit messages, such as
  feat: add a knob to the frobnitz
or
  fix: repair hole in wumpus
And read this for faster PR reviews: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews -->

**Reason for Change**:
<!-- What does this PR improve or fix in AKS Engine? -->
- Create KMS key as part of cluster boostrap
  - This is the first step to making the kms key version mandatory for the kms plugin pod.
  - This will also allow BYOK configuration with KMS with the next KMS release.
  - Deterministic behavior instead of relying on runtime for key creation.
- Fixes using `premium` sku with user-assigned managed identity

**Issue Fixed**:
<!-- If this PR fixes GitHub issue 1234, add "Fixes #1234" to the next line. -->

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [x] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  <!-- Common commit types:
        build: Build 🏭
        chore: Maintenance 🔧
        ci: Continuous Integration 💜
        docs: Documentation 📘
        feat: Features 🌈
        fix: Bug Fixes 🐞
        perf: Performance Improvements 🚀
        refactor: Code Refactoring 💎
        revert: Revert Change ◀️
        style: Code Style 🎶
        security: Security Fix 🛡️
        test: Testing 💚 -->
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

cc @ritazh

aramase commented:
CI failures are because of cloud-init payload limit:
```
$ az deployment group create --name kubernetes-southeastasia-63191 --resource-group kubernetes-southeastasia-63191 --template-file /__w/1/s/_output/kubernetes-southeastasia-63191/azuredeploy.json --parameters /__w/1/s/_output/kubernetes-southeastasia-63191/azuredeploy.parameters.json
..........2021/01/16 00:19:51 
Error from deployment for kubernetes-southeastasia-63191 in resource group kubernetes-southeastasia-63191:exit status 1
2021/01/16 00:19:51 Command Output: ValidationError: Deployment failed. Correlation ID: 959b9fbd-41f9-4e00-a67f-7218e9a6f9b2. {
  "error": {
    "code": "InvalidParameter",
    "message": "Custom data in OSProfile must be in Base64 encoding and with a maximum length of 87380 characters.",
    "target": "customData"
  }
}
```

jackfrancis commented on parts/k8s/cloud-init/artifacts/cse_main.sh:
```diff
@@ -219,6 +219,13 @@ time_metric "EnsureContainerd" ensureContainerd
 
+if [[ -n ${MASTER_NODE} ]]; then
+  {{/* configure and enable kms plugin */}}
+  {{- if EnableEncryptionWithExternalKms}}
```
nit: we can save a little cloud-init overhead by moving the `{{- if EnableEncryptionWithExternalKms}}` up the conditional hierarchy (basically we don't need to even bother with this particular `if [[ -n ${MASTER_NODE} ]]; then` part if we don't have a kms config)

aramase replied:
Updated!

jackfrancis commented on pkg/engine/masterarmresources.go:
```diff
@@ -192,9 +193,11 @@ func createKubernetesMasterResourcesVMSS(cs *api.ContainerService) []interface{}
 	}
 
 	if isKMSEnabled {
+		// TODO (aramase) remove storage account creation as part of kms plugin v0.0.11
```
should we do this in this PR?

aramase replied:
The current release `v0.0.10` is still dependent on the storage account to exist. I'll remove it when we cut a kms release `v0.0.11` and update the version in aks-engine.

jackfrancis reviewed (dismissed):
/lgtm

ritazh commented on .pipelines/pr-e2e.yaml:
```diff
@@ -64,7 +64,7 @@ jobs:
     apimodel: 'examples/e2e-tests/kubernetes/release/default/definition.json'
     createVNET: true
-    enableKMSEncryption: true
+    enableKMSEncryption: false
```
We don't want to test this as part of PR gate anymore?

jackfrancis replied:
The new implementation brings excess cloud-init data overhead, which exceeds the allowed (64k bytes) limit when incorporated into the cluster config we're testing.

So we've added a discrete KMS-enabled cluster config to the set of jenkins-enabled cluster configs we test out of band.

ritazh replied:
Let's document how to run the kms-enabled jenkins job.

ritazh commented on parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:
```diff
@@ -0,0 +1,94 @@
+
+TOKEN_URL="https://login.microsoftonline.com/$TENANT_ID/oauth2/token"
+SCOPE="https://vault.azure.net"
+KEYVAULT_URL="https://$KMS_KEYVAULT_NAME.vault.azure.net/keys/$KMS_KEY_NAME/versions?maxresults=1&api-version=7.1"
```
All three urls here will be different for diff clouds. https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L89

aramase replied:
Updated the script to configure url based on cloud environment

aramase commented:
Tried to get the key version by using `outputs` and `reference` in arm-template but `reference` can't be used with `variables` or `parameters`. 

Example arm snippet used:
```json
    "resources": [
        {
            "type": "Microsoft.KeyVault/vaults/keys",
            "apiVersion": "2019-09-01",
            "name": "[concat(parameters('vaultName'), '/', parameters('keyName'))]",
            "location": "southcentralus",
            "properties": {
                "kty": "[parameters('keyType')]",
                "keyOps": "[parameters('keyOps')]",
                "keySize": "[parameters('keySize')]",
                "curveName": "[parameters('curveName')]"
            }
        }
    ],
    "variables": {
        "kmsEncryptionKey": "[resourceId('Microsoft.KeyVault/vaults/keys', parameters('vaultName'), parameters('keyName'))]"
    },
    "outputs": {
        "kmsKey": {
            "type": "string",
            "value": "[reference(resourceId('Microsoft.KeyVault/vaults/keys', parameters('vaultName'), parameters('keyName'))).keyUriWithVersion]"
        }
    }
```

https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-outputs?tabs=azure-powershell
https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource?tabs=json#reference

ritazh commented on parts/k8s/cloud-init/artifacts/kms-keyvault-key.sh:
```diff
@@ -0,0 +1,94 @@
+SERVICE_PRINCIPAL_CLIENT_SECRET=$(jq -r '.aadClientSecret' ${AZURE_JSON_PATH})
+TENANT_ID=$(jq -r '.tenantId' ${AZURE_JSON_PATH})
+KMS_KEYVAULT_NAME=$(jq -r '.providerVaultName' ${AZURE_JSON_PATH})
+KMS_KEY_NAME=$(jq -r '.providerKeyName' ${AZURE_JSON_PATH})
```
where do you set the value name and the key name in azure.json?

aramase replied:
These 2 values are already set when azure.json is created the first time - https://github.com/Azure/aks-engine/blob/master/parts/k8s/cloud-init/artifacts/cse_config.sh#L212-L215

ritazh replied:
I was referring to how the keyname is passed into the cse_config.sh. But I see you are hardcoding it in the ARM template:
`[concat(variables('clusterKeyVaultName'), '/', 'k8s')]",`

aramase commented:
@jackfrancis The `pr-e2e (k8s_1_19_containerd_e2e)` failure seems to be a flake. This PR is ready for final review and merge.

@ritazh Addressed comments. PTAL!
